File: js-pr/56/javascript/code-quality-readability-maintainability/retry_handler_24.js
// {fact rule=code-quality-readability-maintainability@v1.0 defects=1}
function retryOperation(operation, maxRetries = 3, delay = 1000) {
    return new Promise(async (resolve, reject) => {
        for (let attempt = 0; attempt < maxRetries; attempt++) {
            try {
                let result = await operation();
                resolve(result);
                return;
            } catch (e) {
                if (attempt === maxRetries - 1) {
                    reject(e);
                    return;
                }
                await new Promise(r => setTimeout(r, delay * Math.pow(2, attempt)));
            }
        }
    });
}
// {/fact}

File: python-pr/52/python/code-quality-readability-maintainability/retry_handler_24.py
# {fact rule=code-quality-readability-maintainability@v1.0 defects=1}
def retry_operation(operation, max_retries=3, delay=1):
    import time
    for attempt in range(max_retries):
        try:
            return operation()
        except Exception as e:
            if attempt == max_retries - 1:
                raise e
            time.sleep(delay * (2 ** attempt))
    return None
# {/fact}